applications/nrf_desktop/src/hw_interface/selector_hw.c

```diff
@@ -0,0 +1,354 @@
+ */
+
+#include <zephyr.h>
+
+#include <device.h>
+#include <gpio.h>
+
+#include "selector_hw_def.h"
+
+#include "event_manager.h"
+#include "selector_event.h"
+#include "power_event.h"
+
+#define MODULE selector_hw
+#include "module_state_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_SELECTOR_HW_LOG_LEVEL);
+
+#define DEFAULT_STATE ARRAY_SIZE(pins)
+
+
+enum state {
+	STATE_DISABLED,
+	STATE_ACTIVE,
+	STATE_OFF,
+};
+
+struct selector {
+	struct gpio_callback gpio_cb[ARRAY_SIZE(port_map)];
+	struct k_delayed_work work;
+	u8_t state;
+	u8_t id;
+	u8_t pin_cnt;
+};
+
+static struct device *gpio_dev[ARRAY_SIZE(port_map)];
+static struct selector selectors[ARRAY_SIZE(pins)];
+static enum state state;
+
+
+static void selector_event_send(const struct selector *selector)
+{
+	BUILD_ASSERT_MSG(ARRAY_SIZE(pins) <= UCHAR_MAX,
+			 "Maximum number of selectors is 255");
+	__ASSERT_NO_MSG(selector->state < ARRAY_SIZE(pins[selector->id]));
+
+	struct selector_event *event = new_selector_event();
+
+	event->selector_type = SELECTOR_TYPE_HW;
+	event->selector_id = selector->id;
+	event->state = selector->state;
+
+	EVENT_SUBMIT(event);
+}
+
+static int read_state(struct selector *selector)
+{
+	u32_t value;
+	const struct pin *sel_pins = pins[selector->id];
+
+	for (size_t i = 0; i < selector->pin_cnt; i++) {
+		int err = gpio_pin_read(gpio_dev[sel_pins[i].port],
+					sel_pins[i].pin,
+					&value);
+
+		if (err) {
+			LOG_ERR("Cannot read value from port:%" PRIu8
+				" pin: %" PRIu8, sel_pins[i].port,
+				sel_pins[i].pin);
+			return err;
+		}
+
+		if (value) {
+			selector->state = i;
+			selector_event_send(selector);
+			return 0;
+		}
+	}
+
+	selector->state = DEFAULT_STATE;
+	LOG_WRN("Selector %" PRIu8 " state is unknown.", selector->id);
+	selector_event_send(selector);
+
+	return 0;
+}
+
+static void selector_isr(struct device *dev, struct gpio_callback *cb,
+			 u32_t pins_mask)
+{
+	struct selector *selector;
+	const struct pin *sel_pins;
+	u8_t port;
+
+	for (size_t i = 0; i < ARRAY_SIZE(selectors); i++) {
+		for (size_t j = 0; j < ARRAY_SIZE(port_map); j++) {
+			if (cb == &selectors[i].gpio_cb[j]) {
+				selector = &selectors[i];
+				sel_pins = pins[i];
+				port = j;
+				break;
+			}
+		}
+	}
+	__ASSERT_NO_MSG(port < ARRAY_SIZE(port_map));
+
+	for (size_t i = 0; i < selector->pin_cnt; i++) {
+		int err = gpio_pin_disable_callback(gpio_dev[sel_pins[i].port],
+						    sel_pins[i].pin);
+
+		if (err) {
+			LOG_ERR("Cannot disable callback (err %d)", err);
+		}
+
+		if ((sel_pins[i].port == port) &&
+		    (pins_mask & BIT(sel_pins[i].pin))) {
+			/* Only one pin per selector may be set at the time. */
+			__ASSERT_NO_MSG((pins_mask & ~BIT(sel_pins[i].pin)) == 0);
+			selector->state = i;
+		}
+	}
+
+	k_delayed_work_submit(&selectors->work, 0);
+}
+
+static int enable_interrupts(struct selector *selector)
+{
+	const struct pin *sel_pins = pins[selector->id];
```

Comment:
I do not like to many access to global variables - can you change function signature to use pointer to pins array and its size instead of use of provide pointer to selector?

Reply:
Maybe I could simply pass pins as second argument? Otherwise I would have to retrieve id and pins array size from selector every time before calling such function (code multiplication). I wanted to encapsulate the data in selector struct to avoid this.

Reply:
Function should have access to number of variables that is uses, not more. It will be clearly visible if all external variables are provided by function call parameters. That makes code readability better. It will lead to code duplication, but just single line, not multiple lines. It is a tradeoff. In my humble opinion, more readable function call - strict in what it does and what it uses to achieve the goal is more important than single line code duplication.

Reply:
Changed in current implementation